core/domain/exp_domain.py

```diff
@@ -3519,3 +3519,226 @@ def to_metadata_dict(self):
+        for (state_id, state_name) in state_id_to_names.iteritems():
+            new_state = new_exploration.states[state_name]
+            old_state = old_exploration.states[old_state_id_to_names[state_id]]
+
+            if cls.has_state_changed_significantly(old_state, new_state):
+                # If significant change has happend then treat that state as
+                # new state and assign new state id to that state.
+                new_state_name_to_ids.pop(state_name)
+                new_state_names.append(state_name)
+
+        # Assign new ids to each state in new state names list.
+        latest_state_id_used = model.latest_state_id_used
+        for state_name in sorted(new_state_names):
+            latest_state_id_used += 1
+            new_state_name_to_ids[state_name] = latest_state_id_used
+
+        state_id_map = cls(
+            new_exploration.id, new_exploration.version, new_state_name_to_ids,
+            latest_state_id_used)
+        state_id_map.validate()
+        return state_id_map
+
+    @classmethod
+    def get_state_id_mapping_for_new_exploration(cls, exploration):
+        """Get state name to id mapping for exploration's states.
+
+        Args:
+            exploration: Exploration. New exploration for which state id mapping
+                is to be generated.
+
+        Returns:
+            StateIdMapping. Domain object for state id mapping.
+        """
+        latest_state_id_used = -1
+        state_name_to_ids = {}
+
+        # Assign state id to each state in exploration.
+        for state_name in exploration.states:
+            latest_state_id_used += 1
+            state_name_to_ids[state_name] = latest_state_id_used
+
+        state_id_map = cls(
+            exploration.id, exploration.version, state_name_to_ids,
+            latest_state_id_used)
+        state_id_map.validate()
+        return state_id_map
+
+    @classmethod
+    def get_state_id_mapping_from_model(cls, mapping_model):
+        """Get state name to id mapping from StateIdMappingModel instance.
+
+        Args:
+            mapping_model: StateIdMappingModel. Instance of StateIDMappingModel.
+
+        Returns:
+            StateIdMapping. Domain object for state id mapping.
+        """
+        state_id_map = cls(
+            mapping_model.exploration_id, mapping_model.exploration_version,
+            mapping_model.get_all_state_name_ids(),
+            mapping_model.latest_state_id_used)
+        state_id_map.validate()
+        return state_id_map
+
+    def validate(self):
+        """Validates the state id mapping domain object."""
+        state_ids = self.state_name_to_ids.values()
+        if len(set(state_ids)) != len(state_ids):
+            raise Exception('Assigned state ids should be unique.')
+
+        if not all(state_ids) <= self.latest_state_id_used:
+            raise Exception(
+                'Assigned state ids should be greater then last state id used.')
```

Comment:
You mean "should not be greater than"? (Note that 'than' is misspelled.)

Reply:
Yes, my mistake.